### ISSUE TITLE
`ExecutionResult` should support input as bytes

### ISSUE THREAD

**Issue by lubrst** [robotframework/robotframework#3194]
According to the documentation only paths to xml output files are allowed as input parameters of the `ExecutionResult` class. So there are the following possible enhancements:

1. Strings can also be also used as input parameters, so the documentation could be more precise here.

2.  It would be great if byte strings could also be allowed as input parameters of the `ExecutionResult` class, as in this Python3 example: `result = ExecutionResult(b'xml-string')`.
At the moment the following exception is thrown `robot.errors.DataError: Reading XML source '<in-memory file>' failed: Invalid argument` if a byte string is used.
As I see in the code, the `_open_source_if_necessary` function in the `ETSource` class is responsible for converting input to file objects, so it might be a small change to add support for byte strings.

If these improvements are okay, I can work on a pull request for them.

**Comment by pekkaklarck** [comment]
Do you have a use case to use bytes as paths?

**Comment by pekkaklarck** [comment]
Also, "paths" in the doc actually mean strings. I don't know do actual `pathlib.Path` objects even work. If not, that should be fixed. Should then also look are there other similar APIs that should be fixed.

**Comment by lubrst** [comment]
Yes, I have a use case. I need to read output.xml files from a MongoDB GridFS to analyze them and they are always byte strings. Right now I have to decode the byte strings into strings first, so the function `_open_source_if_necessary` has to encode them back into byte strings. This unnecessary conversion could be avoided by directly supporting byte strings.

In addition, `ExecutionResult` are working with `pathlib.Path` parameters.

**Comment by pekkaklarck** [comment]
OK. Encoding on the client side shouldn't be a big problem, but I don't see problems fixing this either. If I remember correctly, file system APIs in Python 3 still support byte paths so most likely everything should work fine if such paths are just passed to them as-is. I'd assume byte paths work already now on Python 2 and with Python 3 the problem would be due to some other check we do for paths. Did you already debug the problem further?

The fix should contain also unit tests to validate the behavior. It would be great to add tests also to the `pathlib.Path` use case.

**Comment by lubrst** [comment]
Yes, the problem seems to be in the `not is_string(self._source)` condition in the following function:
```    
def _open_source_if_necessary(self):
    if self._source_is_file_name() or not is_string(self._source):
        return None
    if IRONPYTHON_WITH_BROKEN_ETREE:
        return StringIO(self._source)
    return BytesIO(self._source.encode('UTF-8'))
```
- In Python 2 `is_string(b'xml-str') returns True, because the byte-string has type 'str'.

- In Python 3 `is_string(b'xml-str') returns False, because the byte-string has type 'bytes'.

In Python 3 the function always returns None, if `self._source` is a byte-string. To get the same behavior as in Python 2, an additional condition can be added with a `is_byte()` check. 

Sould I make a fix for this behavior and add some new unit tests?

**Comment by pekkaklarck** [comment]
You are most likely right. `is_string` returning False with bytes on Python 3 is by design but in this context bytes ought to be fine. Could use `is_bytes` but I think it considers also `bytearrays` bytes which may not be what we want. Perhaps just plain `isinstance(x, bytes)` would be better.

Unit tests would be good to validate the fix and to make sure the functionality isn't broken in the future.

**Comment by pekkaklarck** [comment]
Reading this conversation again it seems I've misunderstood the original issue. I seem to be talking about paths given to ExecutionResult as bytes, but you apparently meant passing XML content as bytes to it. I didn't even remember it supports XML like that but ETSource enables it.

Anyway, supporting XML as bytes directly is definitely a useful feature and something that our XML library could also use as #3226 demonstrates. In addition to that, it would be good to support paths specified as bytes.

**Comment by pekkaklarck** [comment]
Noticed that the original code didn't work if you specified XML content as a Unicode string so that it used some other encoding than UTF-8. Luckily our own XMLs are UTF-8 and I would expect using some other encoding is pretty rare in general. I got fix to that problem and some cleanup/fixes to PR #3197 (fix for pathlib.Path w/ Python < 3.6, doc update, ...) ready and will commit them shortly.
